feature: Workaround to allow passing an ARN for a Secret in addition to the name

patrickherrera opened the pull request:
ARNs contain a ':' characters, which the upstream parsing logic in Kafka already treats as a separator and will not handle correctly.

This allows you to mask these characters with an alternative (say '|'), which is safely passed through and then converted back before retrieving the Secret.

A new configuration param is exposed (`separator.replacement`) to allow you to specify the replacement character. This will simply be search and replaced later (if the replacement character does not exist in the string, no changes will be made or error thrown).

Fixes aws-samples/msk-config-providers#22

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

patrickherrera committed:
feature: Workaround to allow passing an ARN for a Secret in addition to the name

ARNs contain a ':' characters, which the upstream parsing logic in Kafka already treats as a separator and will not handle correctly.

This allows you to mask these characters with an alternative (say '|'), which is safely passed through and then converted back before retrieving the Secret.

A new configuration param is exposed (`separator.replacement`) to allow you to specify the replacement character. This will simply be search and replaced later (if the replacement character does not exist in the string, no changes will be made or error thrown).

Fixes aws-samples/msk-config-providers#22

patrickherrera commented:
Hi @bdesert, I don't have permission to add you as a reviewer, so here is my PR, thanks.
I struggled coming up with good names and descriptions in a few places, so suggestions are welcome

bdesert commented on pom.xml:
```diff
@@ -2,7 +2,7 @@
 	<groupId>software.amazon.msk</groupId>
 	<artifactId>msk-config-providers</artifactId>
-	<version>0.3.0-SNAPSHOT</version>
+	<version>0.4.0-SNAPSHOT</version>
```
you don't need to change snapshot, I'll change it as part of the version release.

bdesert commented on src/main/java/com/amazonaws/kafka/config/providers/SecretsManagerConfigProvider.java:
```diff
@@ -111,31 +111,30 @@ public ConfigData get(String path) {
 			SecretsManagerClient secretsClient = checkOrInitSecretManagerClient();
-            GetSecretValueResponse response = secretsClient .getSecretValue(request);
+            GetSecretValueResponse response = secretsClient.getSecretValue(request);
 			String value = response.secretString();
```
can you please replace all `tab` chars with 4 spaces?

bdesert commented on src/main/java/com/amazonaws/kafka/config/providers/common/AwsServiceConfigProvider.java:
please review the comments to the issue, I think we won't need to update config portion

bdesert commented on src/main/java/com/amazonaws/kafka/config/providers/common/CommonConfigUtils.java:
please review the comments to the issue, I think we won't need to update config portion

bdesert reviewed (commented):
please review the comments to the issue, let me know what you think about suggested approach

patrickherrera committed:
feature: Support URL-encoded secret paths to allow passing an ARN for a Secret in addition to the name

ARNs contain ':' characters, which the upstream parsing logic in Kafka already treats as a separator and will not handle correctly.

Paths will now be URL-decoded before being looked up in Secrets Manager, so URL-encoding them first (application/x-www-form-urlencoded) will allow them to pass through the initial parser.

Fixes aws-samples/msk-config-providers#22

patrickherrera committed:
feature: Support URL-encoded secret paths to allow passing an ARN for a Secret in addition to the name

ARNs contain ':' characters, which the upstream parsing logic in Kafka already treats as a separator and will not handle correctly.

Paths will now be URL-decoded before being looked up in Secrets Manager, so URL-encoding them first (application/x-www-form-urlencoded) will allow them to pass through the initial parser.

Fixes aws-samples/msk-config-providers#22

patrickherrera commented on README.md:
````diff
@@ -30,8 +30,14 @@ config.providers.s3import.param.region         = us-west-2
 ```
 
+> NOTE: For the `secretsmanager` provider, the base parser from Kafka does not support using ARNs due to the `:` character. 
+> In this case, URL encode the ARN first, and it will be decoded correctly by this Provider. 
````
I decided to keep the docs simpler by recommending only URL-encoding if you need to

patrickherrera commented:
Hi @bdesert, have you had a chance to look at this yet? :-)

bdesert commented:
+1 LGTM
thanks for the contribution (including UT)
Merged
